### PR TITLE
feat(flags): table list view updates

### PR DESCRIPTION
**Pull request by liyiy**
## Changes

small updates to align with the figma mocks. 
- add status/user select filters
- use lemon tags
- wording

Screen Shot 2022-08-31 at 9 39 15 PM

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/api/feature_flag.py [hunk 1]
```diff
@@ -193,6 +194,12 @@ class FeatureFlagViewSet(StructuredViewSetMixin, ForbidDestroyModel, viewsets.Mo
 
     def get_queryset(self) -> QuerySet:
```
Comment (neilkakkar) [inline posthog/api/feature_flag.py:195]: Suggestion: Since flags are user generated, and so far, the max flags in a team is 250, do all this filtering in the frontend itself.

It doesn't make sense to make separate requests to the backend whenever you change filters on the frontend. Will ensure things are much faster.

([Statistic link](https://metabase.posthog.net/question#eyJuYW1lIjpudWxsLCJkYXRhc2V0X3F1ZXJ5Ijp7InF1ZXJ5Ijp7InNvdXJjZS10YWJsZSI6NDcsImFnZ3JlZ2F0aW9uIjpbWyJjb3VudCJdXSwiYnJlYWtvdXQiOltbImZpZWxkIiwzNzksbnVsbF1dLCJvcmRlci1ieSI6W1siZGVzYyIsWyJhZ2dyZWdhdGlvbiIsMF1dXX0sInR5cGUiOiJxdWVyeSIsImRhdGFiYXNlIjozfSwiZGlzcGxheSI6InRhYmxlIiwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319) )

Comment (liyiy) [reply posthog/api/feature_flag.py:195]: I was doing that but it's much cleaner to do it in the backend as the logic gets pretty wieldy to work with otherwise, plus it doesn't target something crazy like clickhouse events or anything right? I don't think users will be playing with those filters too often anyway so a couple of requests to feature_flags seems okay unless there are other implications?

Comment (neilkakkar) [reply posthog/api/feature_flag.py:195]: Elaborate a bit on what's getting unwieldy?

This introduces a place where it _can_ become a future problem: When someone starts complaining about it being slow, and then we end up fixing it.

It doesn't _seem_ like it should be too unwieldy logic wise: Having a selector that takes in `featureFlags`, and `filters`, which does a filter on the two possible keys; and then using this selector in the list instead of `featureFlags` ? What am I missing?

Comment (liyiy) [reply posthog/api/feature_flag.py:195]: It also gives a more accurate reflection of state than simply filtering it out on the frontend if someone else were to create/edit/delete a new feature flag or enable/disable flags in that time. All of our page filtering mechanisms for persons, dashboards, events, experiments, pretty much use the backend to filter too and even if it seems "unnecessary" for now to use the backend to filter for two feature flag options, wouldn't it be better to stay with convention if the benefits to using the frontend don't really outweigh it for the backend ? It's also more extensible if we use the backend to filter, like if we were to add more filtering capabilities, no?

> When someone starts complaining about it being slow, and then we end up fixing it.
Why would it become slow? Can you give an example?

Comment (liyiy) [reply posthog/api/feature_flag.py:195]: If someone is on a slow 3G connection wouldn't the app be pretty slow for them in general? 🤔  I'm happy to refactor this into the frontend if this ever crops up as an issue specifically for feature flags + filtering 🤗 but I don't see anything wrong with using the backend for now as that was a conscious decision made to do so 👍

Comment (neilkakkar) [reply posthog/api/feature_flag.py:195]: >If someone is on a slow 3G connection wouldn't the app be pretty slow for them in general?

>> It's not just for slow 3G connections, but people away from our servers, where latency can be high, things add up.

**Review by clarkus — CHANGES_REQUESTED** [review]
The created column of time values should be right-aligned to make it easier to compare values. Otherwise looks good to me.

**Comment by liyiy** [comment]
> The created column of time values should be right-aligned to make it easier to compare values. Otherwise looks good to me.

more right aligned now for easy comparison and also sorting based on creation time @neilkakkar 

Screen Shot 2022-09-01 at 4 10 48 PM

**Comment by clarkus** [comment]
> > The created column of time values should be right-aligned to make it easier to compare values. Otherwise looks good to me.
> 
> more right aligned now for easy comparison and also sorting based on creation time @neilkakkar
> 
> Screen Shot 2022-09-01 at 4 10 48 PM

This doesn't look right-aligned - the values should all share the same right edge / offset. The header for the column should also be right aligned with the values in the column. 

Screen Shot 2022-09-01 at 2 02 47 PM

**Review by clarkus — CHANGES_REQUESTED** [review]
This doesn't look right-aligned - the values should all share the same right edge / offset. The header for the column should also be right aligned with the values in the column. 

Screen Shot 2022-09-01 at 2 02 47 PM

File: frontend/src/scenes/feature-flags/FeatureFlags.tsx [hunk 2]
```diff
@@ -68,14 +68,13 @@ function OverViewTab(): JSX.Element {
...
-                        }
-                        label={<span className="font-normal">{featureFlag.active ? 'Enabled' : 'Disabled'}</span>}
-                    />
+                    <>
+                        {featureFlag.active ? (
+                            <LemonTag type="success">Enabled</LemonTag>
+                        ) : (
+                            <LemonTag type="default">Disabled</LemonTag>
+                        )}
+                    </>
```
Comment (Twixes) [inline frontend/src/scenes/feature-flags/FeatureFlags.tsx:77]: I wonder what's the motivation behind replacing the interactive switch with a static tag?

Comment (liyiy) [reply frontend/src/scenes/feature-flags/FeatureFlags.tsx:77]: > This doesn't look right-aligned - the values should all share the same right edge / offset. The header for the column should also be right aligned with the values in the column.
> 
> Screen Shot 2022-09-01 at 2 02 47 PM

The "created by" column is a reuseable column component, if I make the change it'll be all across the board, did you want that or did you just want it to be for feature flags specifically?

**Comment by liyiy** [comment]
> This doesn't look right-aligned - the values should all share the same right edge / offset. The header for the column should also be right aligned with the values in the column.
> 
> Screen Shot 2022-09-01 at 2 02 47 PM

The "created by" column is a reuseable column component, if I make the change it'll be all across the board, did you want that or did you just want it to be for feature flags specifically?

File: frontend/src/scenes/feature-flags/FeatureFlags.tsx (hunk 2 above)
Comment (liyiy) [reply frontend/src/scenes/feature-flags/FeatureFlags.tsx:77]: For more context we replace it with a more actions dropdown option of enabling/disabling the flag but this one is for @clarkus

Comment (clarkus) [reply frontend/src/scenes/feature-flags/FeatureFlags.tsx:77]: That action has moved to an inline menu. You can see the full designs at https://www.figma.com/file/eX6AKgqX2bZDxgEg2S4xzy/Feature-Flags-%26-Experiments?node-id=0%3A1

**Comment by clarkus** [comment]
> The "created by" column is a reuseable column component, if I make the change it'll be all across the board, did you want that or did you just want it to be for feature flags specifically?

Numeric columns in tables should almost always be right aligned. In this specific case, it should be right aligned and generally I think it's safe to do this for all created by columns that contain relative date values.

File: frontend/src/lib/components/LemonTable/columnUtils.tsx [hunk 3]
```diff
@@ -7,11 +7,11 @@ import { UserBasicType } from '~/types'
 
 export function createdAtColumn<T extends { created_at?: string | null }>(): LemonTableColumn<T, 'created_at'> {
     return {
-        title: 'Created on',
+        title: 'Created',
         dataIndex: 'created_at',
         render: function RenderCreatedAt(created_at) {
             return created_at ? (
-                <div style={{ whiteSpace: 'nowrap' }}>
+                <div style={{ whiteSpace: 'nowrap', textAlign: 'center' }}>
```
Comment (clarkus) [inline frontend/src/lib/components/LemonTable/columnUtils.tsx:14]: Why was this changed to centered?

Comment (liyiy) [reply frontend/src/lib/components/LemonTable/columnUtils.tsx:14]: because it looks really horrible when it's right aligned completely as there's extra padding/considerations that'll be needed to offset it

Screen Shot 2022-09-02 at 12 03 03 PM

Comment (mariusandra) [reply frontend/src/lib/components/LemonTable/columnUtils.tsx:14]: flyby: I think if both the header and column are right-aligned, it'll look just fine...

Comment (clarkus) [reply frontend/src/lib/components/LemonTable/columnUtils.tsx:14]: Yeah for it to work the column header and contents should all have the same alignment. 
Screen Shot 2022-09-02 at 9 16 57 AM

**Comment by liyiy** [comment]
@clarkus 
Screen Shot 2022-09-02 at 12 37 09 PM

**Review by clarkus — APPROVED** [review]

File: frontend/src/scenes/feature-flags/featureFlagsLogic.ts [hunk 4]
```diff
@@ -60,6 +71,24 @@ export const featureFlagsLogic = kea<featureFlagsLogicType>({
                 },
             ],
         ],
+        uniqueCreators: [
+            (selectors) => [selectors.featureFlags],
+            (featureFlags) => {
+                const creators = {}
+                featureFlags.forEach((flag: FeatureFlagType) => {
```
Comment (mariusandra) [inline frontend/src/scenes/feature-flags/featureFlagsLogic.ts:78]: I don't think you need to type `: FeatureFlagType` here. It should come directly from the input selector. 

I'd personally also write this as a simple `for (const flag of featureFlags) {` instead of `forEach`. Mostly because it's a [bit](https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/) faster and I think also cleaner with scoping... The real world speed difference for one loop really negligible, so treat this as a nit.

File: frontend/src/scenes/feature-flags/featureFlagsLogic.ts [hunk 5]
```diff
@@ -60,6 +71,24 @@ export const featureFlagsLogic = kea<featureFlagsLogicType>({
...
+                        if (!creators[flag.created_by.id]) {
+                            creators[flag.created_by.id] = flag.created_by.first_name
+                        }
+                    }
+                })
+                const response: LemonSelectOption<string>[] = [{ label: 'Any user', value: 'any' }]
+                for (const [id, first_name] of Object.entries(creators)) {
+                    response.push({ label: first_name, value: id } as LemonSelectOption<string>)
+                }
+                return response
```
Comment (mariusandra) [inline frontend/src/scenes/feature-flags/featureFlagsLogic.ts:89]: Specifying ` as LemonSelectOption<string>` can hide subtle bugs. TypeScript should be smart enough to infer what type the data you're pushing to `response` should be.

Normally you shouldn't need to type any variable more than once.

Also, if you want to get fancy, here's another way to do this:

```ts
const response: LemonSelectOption<string>[] = [
    { label: 'Any user', value: 'any' },
    ...(Object.entries(creators).map(
        ([id, first_name]) => ({ label: first_name, value: id }),
    )),
]
```

File: frontend/src/lib/components/LemonTable/columnUtils.tsx [hunk 6]
```diff
@@ -7,17 +7,18 @@ import { UserBasicType } from '~/types'
 
 export function createdAtColumn<T extends { created_at?: string | null }>(): LemonTableColumn<T, 'created_at'> {
     return {
-        title: 'Created on',
+        title: 'Created',
         dataIndex: 'created_at',
         render: function RenderCreatedAt(created_at) {
             return created_at ? (
-                <div style={{ whiteSpace: 'nowrap' }}>
+                <div style={{ whiteSpace: 'nowrap', textAlign: 'right' }}>
```
Comment (mariusandra) [inline frontend/src/lib/components/LemonTable/columnUtils.tsx:14]: Whenever you feel yourself reaching to add anything to a style tag, see if you can refactor it with utility classes instead:

```suggestion
                <div className='whitespace-nowrap text-right'>
```

**Review by mariusandra — APPROVED** [review]
Code looks good too, left some nits, nothing blocking.

**Comment by mariusandra** [comment]
Jest is running into some issues though. I think a merge from master should fix that.

File: frontend/src/lib/components/LemonTable/columnUtils.tsx (hunk 6 above)
Comment (liyiy) [reply frontend/src/lib/components/LemonTable/columnUtils.tsx:14]: got it 🤗
